### ISSUE TITLE
sequelize.sync not completed when resolving

### ISSUE THREAD

**Issue by tgroutars** [sequelize/sequelize#8918]
## What are you doing?
I'm using Jest to test my code. I use a setup script with the following code

```js
const { sequelize } = require('/path/to/models');

beforeAll(async () => {
  await sequelize.sync();
});
```
However, everytime I add a new model, my tests fail once on circleci, then get fixed when I rebuild. I'm pretty sure sequelize.sync hasn't finished at the time it resolves.

As a workaround, I instantiate a new instance of Sequelize to perform the sync and close it before running my tests

```js
const { 
  sequelize, 
  newSequelize /* instantiates a new instance with the same config */ 
} = require('/path/to/models');

beforeAll(async () => {
  const setupSequelize = newSequelize();
  await setupSequelize.sync();
  await setupSequelize.close();
});
```
But it seems hacky

## What do you expect to happen?
I expect sequelize.sync to have completed at the time it resolves

## What is actually happening?
sequelize.sync resolves before having completed

__Dialect:__ postgres
__Dialect version:__ pg@6.4.2
__Database version:__ PostgreSQL 10.1
__Sequelize version:__ sequelize@4.31.0
__Tested with latest release:__ No, but it's not easy to reproduce as the error doesn't happen all the time, only when the tests start before sequelize.sync having completed

**Comment by sushantdhiman** [comment]
`sync` does not have any non returned promises. A short tests case

**JS**
```js
const Model = sequelize.define('Model', {
  name: Sequelize.STRING
});

async function test () {
  console.log('before');

  await Model.sync();

  console.log('after');
}

test();
```
**OUTPUT**
```sql
before
Executing (default): CREATE TABLE IF NOT EXISTS "Models" ("id"   SERIAL , "name" VARCHAR(255), "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL, "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL, PRIMARY KEY ("id"));
Executing (default): SELECT i.relname AS name, ix.indisprimary AS primary, ix.indisunique AS unique, ix.indkey AS indkey, array_agg(a.attnum) as column_indexes, array_agg(a.attname) AS column_names, pg_get_indexdef(ix.indexrelid) AS definition FROM pg_class t, pg_class i, pg_index ix, pg_attribute a WHERE t.oid = ix.indrelid AND i.oid = ix.indexrelid AND a.attrelid = t.oid AND t.relkind = 'r' and t.relname = 'Models' GROUP BY i.relname, ix.indexrelid, ix.indisprimary, ix.indisunique, ix.indkey ORDER BY i.relname;
after
```

Are you sure this is not something related to your setup, may be missing promise return somewhere. Because our own tests are quite dependent on `sync` and never heard about such issue before.

**Comment by tgroutars** [comment]
Well that's the thing that makes it pretty hard to reproduce. Just because it does resolve before the sync has completed doesn't mean the test will always start running before it's completed. 

It might be another issue linked to my test config, but it's pretty straightforward. Here are my logs from circleci:

```
yarn run v1.3.2
$ cross-env NODE_ENV=test jest --config test/server/jest.config.json
 FAIL  test/server/models/user.test.js (7.054s)
  ● User › #validPassword › returns false if the password is wrong

    SequelizeDatabaseError: relation "slack_workspace" does not exist
      
      at Query.formatError (../../node_modules/sequelize/lib/dialects/postgres/query.js:356:16)
      at query.catch.err (../../node_modules/sequelize/lib/dialects/postgres/query.js:86:18)
      at tryCatcher (../../node_modules/bluebird/js/release/util.js:16:23)
      at Promise._settlePromiseFromHandler (../../node_modules/bluebird/js/release/promise.js:512:31)
      at Promise._settlePromise (../../node_modules/bluebird/js/release/promise.js:569:18)
      at Promise._settlePromise0 (../../node_modules/bluebird/js/release/promise.js:614:10)
      at Promise._settlePromises (../../node_modules/bluebird/js/release/promise.js:689:18)
      at Async.Object.<anonymous>.Async._drainQueue (../../node_modules/bluebird/js/release/async.js:133:16)
      at Async.Object.<anonymous>.Async._drainQueues (../../node_modules/bluebird/js/release/async.js:143:10)
      at Immediate.Async.drainQueues (../../node_modules/bluebird/js/release/async.js:17:14)
      at runCallback (../../timers.js:789:20)
      at tryOnImmediate (../../timers.js:751:5)
      at processImmediate [as _immediateCallback] (../../timers.js:722:5)

  ● User › #validPassword › returns false if the password is wrong

    SequelizeUniqueConstraintError: Validation error
      
      at Query.formatError (../../node_modules/sequelize/lib/dialects/postgres/query.js:318:18)
      at query.catch.err (../../node_modules/sequelize/lib/dialects/postgres/query.js:86:18)
      at tryCatcher (../../node_modules/bluebird/js/release/util.js:16:23)
      at Promise._settlePromiseFromHandler (../../node_modules/bluebird/js/release/promise.js:512:31)
      at Promise._settlePromise (../../node_modules/bluebird/js/release/promise.js:569:18)
      at Promise._settlePromise0 (../../node_modules/bluebird/js/release/promise.js:614:10)
      at Promise._settlePromises (../../node_modules/bluebird/js/release/promise.js:689:18)
      at Async.Object.<anonymous>.Async._drainQueue (../../node_modules/bluebird/js/release/async.js:133:16)
      at Async.Object.<anonymous>.Async._drainQueues (../../node_modules/bluebird/js/release/async.js:143:10)
      at Immediate.Async.drainQueues (../../node_modules/bluebird/js/release/async.js:17:14)
      at runCallback (../../timers.js:789:20)
      at tryOnImmediate (../../timers.js:751:5)
      at processImmediate [as _immediateCallback] (../../timers.js:722:5)

 PASS  test/server/router/api/auth.test.js (22.074s)

Test Suites: 1 failed, 1 passed, 2 total
Tests:       1 failed, 18 passed, 19 total
Snapshots:   0 total
Time:        22.71s
Ran all test suites.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Exited with code 1
```

My jest config:

```json
{
  "testEnvironment": "node",
  "modulePaths": [
    "../../server"
  ],
  "setupTestFrameworkScriptFile": "./setup.js"
}
```

And my setup script
```js
const { sequelize } = require('models');
const redisClient = require('lib/redisClient');
const { cleanDb } = require('./testUtils');

beforeAll(async () => {
  await sequelize.query('CREATE EXTENSION IF NOT EXISTS "uuid-ossp"');
  await sequelize.sync({ force: true });
});

beforeEach(() => cleanDb());

afterAll(() => sequelize.close());
afterAll(() => redisClient.quit());
```

And the cleanDB method:

```
const { sequelize } = require('models');

const cleanDb = async () => {
  const tableNames = Object.keys(sequelize.models);
  await sequelize.query(
    `TRUNCATE ${tableNames.map(name => `"${name}"`).join(', ')} CASCADE;`
  );
};

module.exports = {
  cleanDb
};
```

The problem might not be sequelize.sync, mind you. But as I see it, it seems like the most plausible cause for the issue. The tests pass when I just rebuild it in the exact same state

**Comment by Jadex1** [comment]
I'm currently using Jest. I don't have this exact setup(I haven't put it on CircleCI yet), but I am using jest. I have noticed that despite having sync and alter on, when jest re runs the tests it doesn't seem to rebuild the db correctly. If I restart the tests and run my whole script it again, everything runs correctly.

**Comment by tgroutars** [comment]
I'm also still experiencing this issue.

Actually it's not just on circleci anymore, I have the exact same problem in my local environment since my app is growing and has more models.

When I re-run the test, there's no problem since the tables already exist, but if I drop and recreate the test db before running my tests, it fails almost every time.

My app is getting bigger, and the more models I have, the more chance I have to stumble upon this issue, which is now almost every time on circleci.

I'm pretty sure the sync method does not finish before resolving, otherwise this would not happen

I'm also pretty sure I don't forget to return a promise anywhere, since the only place I could forget it is in jest's beforeAll

@sushantdhiman is there any news about this ?

**Comment by jackharich** [comment]
I discovered today I have this issue too. sync sometimes behaves properly when handled as a promise, and sometimes it does not. 

Specifically, examine this terminal readout:
[nodemon] starting `node ./work/CsfServer.js`
Server ready at http://localhost:3000
Begin add table to database: countries
   Complete add table to database: countries
Begin add table to database: organizations
Table already in database: users
Add initial records to countries
Add initial records to organizations
Unhandled rejection SequelizeDatabaseError: Table 'thwink_truth1.organizations' doesn't exist
    at Query.formatError (C:\AWork\PPKB\Gen2\Dev\trs_current_git\node_modules\sequelize\lib\dialects\mysql\query.js:247:16)
    at Query.handler [as onResult] (C:\AWork\PPKB\Gen2\Dev\trs_current_git\node_modules\sequelize\lib\dialects\mysql\query.js:68:23)
    at Query.Command.execute (C:\AWork\PPKB\Gen2\Dev\trs_current_git\node_modules\mysql2\lib\commands\command.js:30:12)
    at Connection.handlePacket (C:\AWork\PPKB\Gen2\Dev\trs_current_git\node_modules\mysql2\lib\connection.js:502:28)
    at PacketParser.onPacket (C:\AWork\PPKB\Gen2\Dev\trs_current_git\node_modules\mysql2\lib\connection.js:81:16)
    at PacketParser.executeStart (C:\AWork\PPKB\Gen2\Dev\trs_current_git\node_modules\mysql2\lib\packet_parser.js:77:14)
    at Socket.<anonymous> (C:\AWork\PPKB\Gen2\Dev\trs_current_git\node_modules\mysql2\lib\connection.js:89:29)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:279:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at TCP.onread (net.js:636:20)

   Complete add table to database: organizations

The error occurs because putting sync in an await failed. The line "Complete add table to database: organizations" should appear immediately after "Begin add table to database: organizations". This would indicate the add completed. All add tables must complete before adding records, due to Sequealize associations.

**Comment by HenryQW** [comment]
I ran into the same error, however it used to work, odd.

Setup:
```js
before(async () => {
  await db.sequelize.sync();
});

describe('database init', () => {
  it('SiteRules should have all pre-defined rules', async () => {
    const siteRule = await db.SiteRule.findAll();
    expect(siteRule.length).to.equal(siteRules.rules.length);
  });

  it('Users should have api key', async () => {
    const user = await db.User.findById(1);
    expect(user.apiKey).to.equal(process.env.KEY);
  });
});
```

Result:
```
  database init
{ status: 'success', message: 'Inserted SiteRule 1: ' }
{ status: 'success', message: 'Inserted SiteRule 2: ' }
{ status: 'success', message: 'Inserted SiteRule 3: ' }
{ status: 'success', message: 'Inserted SiteRule 4: ' }
{ status: 'success', message: 'Inserted SiteRule 5: ' }
{ status: 'success', message: 'Inserted SiteRule 6: ' }
{ status: 'success', message: 'Inserted SiteRule 7: ' }
    1) SiteRules should have all pre-defined rules
{ status: 'success', message: 'Inserted SiteRule 8: ' }
{ status: 'success', message: 'Inserted SiteRule 9: ' }
{ status: 'success', message: 'Inserted SiteRule 10: ' }
    ✓ Users should have api key


  1 passing (2s)
  1 failing

  1) database init
       SiteRules should have all pre-defined rules:

      AssertionError: expected 7 to equal 10
      + expected - actual

      -7
      +10
```

**Comment by nhkhanh** [comment]
Sequelize keeps connections to database. You should close Sequelize connections with `sequelize.close();` on after/afterall so Mocha/Jest can exit.

**Comment by kedwin** [comment]
the problem is when you have 2 test files,
jest will run them concurrently sometimes
so if you are using the same database
and have beforeAll in both test files with `db.sync({force:yes });` example

it will break sometimes

[here you can read it](https://github.com/facebook/jest/issues/5818#issuecomment-375927362)

**Comment by m-mitchell** [comment]
There is an open bug in Jest that can make it look like `sequelize.sync()` has resolved when really it's just timed-out partway through: https://github.com/facebook/jest/issues/6619 It can't explain all the issues in this thread but may explain some of them.

**Comment by wildcookie007** [comment]
I ran into this as well. I was running `sql.sync` with force flag in the index of my exported sequelize config and was only forcing if the environment is `test`. But the database never fully synced and omitted my seeding function which came immediately after sync. The only solution that worked for no appearent reason yet is this:
`
        beforeAll(async () => {
		await sql.sync({force: true, logging: false});
		await SeedTestDB();
	});

	afterAll(async () => {
		await sql.drop();
	});
`

**Comment by mtn** [comment]
Did anyone get this deterministically working? Running into the same problem and I've tried a few of the notes above.

**Comment by wildcookie007** [comment]
> Did anyone get this deterministically working? Running into the same problem and I've tried a few of the notes above.

Yes, you have to make sure the tests are running sequentially and not in parallel. Use `--runInBand` flag.

**Comment by mtn** [comment]
> > Did anyone get this deterministically working? Running into the same problem and I've tried a few of the notes above.
> 
> Yes, you have to make sure the tests are running sequentially and not in parallel. Use `--runInBand` flag.

Hm, that seems not to be my problem. With a single test, even including a `await .sync()` in `beforeAll` or elsewhere in the testing harness, I always end up with undefined table errors. But if I move the `sync` outside the main codepath, it works.

**Comment by outluch** [comment]
Investigated in my case:
I had `mocha -w` to tdd. And it was running sync in tests while app was running sync too! My real source code. So it is collision of syncs :)

**Comment by ValheKouneli** [comment]
I had a similar problem using the mocha test library. Test failed the first time on Docker Hub, but when I re-build, all passed. I did not have to update the models for this to happen.

I also used sequelize.sync({ force: true }) on my afterEach hook. When I looked at the logs on Docker Hub, I found out that my afterEach hook occasionally timed-out. After that, the immediately following test failed ( "SequelizeDatabaseError: relation "[model name]" does not exist "). Each time this happened, it was different two tests.

I increased the timeout of my afterEach hook significantly to solve the problem. It was obvious that the following test would fail if the preceding synchronization was interrupted by a time-out. If your afterEach hook does not time-out, this probably won't help.

While researching this problem, I also found out that if I set the timeout of my afterEach hook to infinity (0) and remove the done() at the end, the tests will just stop after the afterEach and PASS. O__O

**Comment by briantical** [comment]
I had defined a `sequelize.sync()` in **models/index.js**. It seemed to be conflicting with `db.sequelize.sync()` in the beforeAll in **setupTests.js**. Restricting the `sequelize.sync()` to run only in the development environment removed the errors.
